includes/class-sensei-analysis-overview-list-table.php

```diff
@@ -28,7 +30,8 @@ class Sensei_Analysis_Overview_List_Table extends Sensei_List_Table {
 	 * @return  void
 	 */
 	public function __construct( $type = 'users' ) {
-		$this->type = in_array( $type, array( 'courses', 'lessons', 'users' ) ) ? $type : 'users';
+		$this->type      = in_array( $type, array( 'courses', 'lessons', 'users' ) ) ? $type : 'users';
+		$this->page_slug = Sensei()->analysis->page_slug;
```

Comment:
Have you considered extracting the page slug into a constant in the `Sensei_Analysis` class?

Reply:
@merkushin I agree, it would be better to be a constant, but I'm not sure how to go about deprecating the public `page_slug` property. Do you think it's worth it to deprecate it or just leave it as is?

Reply:
I'm not actually sure how we can deprecate properties. Maybe using the class magic methods?

```php
public function __get( $key ) {
	if ( 'page_slug' === $key ) {
		_doing_it_wrong( ... );

		return self::PAGE_SLUG;
	}
}
```

Reply:
Sorry, forgot to answer here. Actually, I don't think we have to deprecate and remove the property.
We can set its value using the new constant.

Ideally, from my perspective, would pass this value through the constructor and set it to `page_slud`. This way makes it easier to test classes with that property (as for any other dependency).
However, I see it requires a lot of work. So we might consider it as an additional task for future improvement.

Reply:
> Actually, I don't think we have to deprecate and remove the property. We can set its value using the new constant.

Yeah, that will work for the `Sensei_Analysis_[type]_List_Table` classes, but where should we put the constant? I guess the most logical host is `Sensei_Analysis`? If we put it there, that class also has a `page_slug` property that I think has to be deprecated to avoid confusion. And if we do it for `page_slug` it makes sense to do it for all the other static properties like `post_type`, `name`, etc.

> Ideally, from my perspective, would pass this value through the constructor

This also would be a bit tricky, because of how the `Sensei_Analysis_[type]_List_Table` [factory](https://github.com/Automattic/sensei/blob/add/rename-analysis-to-reports/includes/class-sensei-analysis.php#L116) works. Maybe it would be easier to do it with a `set_page_slug` setter function?

With all of the above said, I'm all up for refactoring this part, but I think I would leave it for some other PR. Maybe when we tackle the report performance issues or improve the test coverage.

Let me know if you see some quick wins here or we are OK with leaving it out for now. 🙂

Reply:
> With all of the above said, I'm all up for refactoring this part, but I think I would leave it for some other PR. Maybe when we tackle the report performance issues or improve the test coverage.

Totally agree that we should leave that work for the future. Too many questions here, most of them are not so easy :)

> I guess the most logical host is Sensei_Analysis?

Yes, I have the same sense.

> If we put it there, that class also has a page_slug property that I think has to be deprecated to avoid confusion. 

It looks like we need to change its visibility to stricter (private), at least.

> And if we do it for page_slug it makes sense to do it for all the other static properties like post_type, name, etc.

Generally, yes. We have a lot of duplication for things like `post_type` in code. It could lead to forgetting to update the value sometimes or inconsistencies. 

Having (as an example) post type in a single place reduces the possibility of accidental mistyping or error. (It doesn't mean that we don't pass the value to other classes and functions, but we keep origin in a single place.)

> Let me know if you see some quick wins here or we are OK with leaving it as is for now. 🙂

That's the toughest point :) I think we can leave as is having in mind that we will come back soon to make it better. (Ideally, create an issue for that.)
However, I should say what is most doubtful for me here.
```
Sensei()->analysis->page_slug
```
This expression creates a strong hidden dependency (because it is not explicitly passed through the class's API) on two outer "things": `Sensei()` function and `Sensei_Analysis` class. And, actually, they pull a bunch of other dependencies under the hood.

How bad is it? I think it is pretty common in our codebase :) But I still don't think it is good, and it would be better to improve our code structure.

How does this improving might help us? It reduces mental load and makes (pure) unit tests possible (and much faster).

Now, as for a quick win. We can introduce the public constant in `Sensei_Analysis` and use its value directly in other places. From my perspective, that's better than that expression.

However, feel free to merge PR without making that. As I can see, we can spend a lot of time just discussing this topic :)

Reply:
Yeah, I agree with all you said. 🙂 You are right that we should try to decouple the code as much as possible moving forward. 

> Now, as for a quick win. We can introduce the public constant in Sensei_Analysis and use its value directly in other places. From my perspective, that's better than that expression.

Done! I've deprecated the `Sensei_Analysis->page_slug` in 908615fe5175ca1ed6d27ea166202db6b57d938c and replace it with a constant. The deprecated property still works, so it shouldn't be causing any issues.

@merkushin Please, give it a look, and if good, merge if this is a blocker for you.